dsgrid/config/time_dimension_base_config.py

```diff
@@ -110,3 +122,78 @@ def list_expected_dataset_timestamps(self):
             List of tuples of columns representing time in the load_data table.
 
         """
+
+    def _convert_time_to_project_time_interval(
+        self, df=None, project_time_dim=None, wrap_time=True
+    ):
+        """Shift time to match project time based on TimeIntervalType
+        If time range spills over into anothr year after time interval alignment,
+        the time range will be wrapped around so it's bounded within the year
+        """
+        if project_time_dim is None:
+            return df
+
+        dtime_interval = self.get_time_interval_type()
+        ptime_interval = project_time_dim.get_time_interval_type()
+        time_col = project_time_dim.get_timestamp_load_data_columns()
+
+        assert len(time_col) == 1, time_col
+        time_col = time_col[0]
+
+        if dtime_interval != ptime_interval:
+            match (dtime_interval, ptime_interval):
+                case (TimeIntervalType.PERIOD_BEGINNING, TimeIntervalType.PERIOD_ENDING):
+                    df = df.withColumn(
+                        time_col,
+                        F.col(time_col)
+                        + F.expr(f"INTERVAL {self.get_frequency().seconds} SECONDS"),
+                    )
+                case (TimeIntervalType.PERIOD_ENDING, TimeIntervalType.PERIOD_BEGINNING):
+                    df = df.withColumn(
+                        time_col,
+                        F.col(time_col)
+                        - F.expr(f"INTERVAL {self.get_frequency().seconds} SECONDS"),
+                    )
+
+        if not wrap_time:
+            return df
+
+        # If dataset_time does not match project_time, try time-wrapping
+        dataset_time = df.agg(F.collect_set(time_col)).collect()[0][0]
+        project_time = (
+            project_time_dim.build_time_dataframe().agg(F.collect_set(time_col)).collect()[0][0]
+        )
+
+        diff = set(dataset_time).difference(set(project_time))
+
+        if diff:
+            # take most common year from time col
+            main_year = (
+                df.groupBy(F.year(time_col)).count().select(f"year({time_col})").collect()[0][0]
+            )
+
+            # time-wrap by changing the year
+            df = (
+                df.filter(F.col(time_col).isin(diff))
+                .withColumn(
+                    time_col,
+                    F.from_unixtime(
+                        F.unix_timestamp(time_col)
+                        + F.datediff(
+                            F.to_date(F.lit(main_year + 1), "yyyy"),
+                            F.to_date(F.year(time_col) + 1, "yyyy"),
+                        )
+                        * 24
+                        * 3600
+                    ).cast("timestamp"),
+                )
+                .union(df.filter(~F.col(time_col).isin(diff)))
+            )
+
+        dataset_time = df.agg(F.collect_set(time_col)).collect()[0][0]
+        diff = set(dataset_time).difference(set(project_time))
```

Comment:
Not a complaint - I'm sure I've written identical lines many times. In case it helps readability, this could be
```
if not dataset_time.issubset(project_time):
    assert/raise
```
It's better to capture the diff if you log it, but you aren't doing that.